drivers/SmartThings/matter-rvc/src/init.lua

```diff
@@ -28,213 +25,232 @@ if version.api < 10 then
+  if current_tag == 0xFFFF then
+    return
+  end
 
-  if is_idle_current_mode then
-    return labels
+  -- Check whether non-idle mode can be selected or not
+  local op_state = capabilities.robotCleanerOperatingState.operatingState
+  local can_be_selected_as_non_idle = 0
+  if current_tag == clusters.RvcRunMode.types.ModeTag.IDLE and
+    (operating_state == op_state.stopped.NAME or operating_state == op_state.paused.NAME or
+     operating_state == op_state.docked.NAME or operating_state == op_state.charging.NAME) then
+      can_be_selected_as_non_idle = 1
   end
 
-  local filtered_labels = {}
-  table.insert(filtered_labels, current_mode)
-  for i, label in ipairs(labels) do
-    if is_idle_mode(device, field_prefix, i, idle_mode_tag) then
-      table.insert(filtered_labels, label)
+  -- Set supported run arguments
+  local supported_arguments = {} -- For generic plugin
+  for i, mode in ipairs(supported_run_modes) do
+    if mode[2] == clusters.RvcRunMode.types.ModeTag.IDLE or can_be_selected_as_non_idle == 1 then
+      table.insert(supported_arguments, mode[1])
     end
   end
-  return filtered_labels
+
+  -- Send event to set supported run arguments
+  local component = device.profile.components["runMode"]
+  local event = capabilities.mode.supportedArguments(supported_arguments, {visibility = {displayed = false}})
+  device:emit_component_event(component, event)
+
+  -- Set supported clean arguments
+  local supported_clean_modes = device:get_field(CLEAN_MODE_SUPPORTED_MODES) or {}
+  supported_arguments = {}
+  for i, mode in ipairs(supported_clean_modes) do
+    table.insert(supported_arguments, mode)
+  end
+
+  -- Send event to set supported clean modes
+  local component = device.profile.components["cleanMode"]
+  local event = capabilities.mode.supportedArguments(supported_arguments, {visibility = {displayed = false}})
+  device:emit_component_event(component, event)
 end
 
 -- Matter Handlers --
-local function rvc_run_mode_supported_mode_attr_handler(driver, device, ib, response)
-  set_field_supported_modes(device, RVC_RUN_MODE_SUPPORTED_MODES, ib.data.elements)
+local function run_mode_supported_mode_handler(driver, device, ib, response)
+  local supported_modes = {}
+  local supported_modes_with_tag = {}
+  for i, mode in ipairs(ib.data.elements) do
+    if version.api < 10 then
+      clusters.RvcRunMode.types.ModeOptionStruct:augment_type(mode)
+    end
+    local tag = 0xFFFF
+    for i, t in ipairs(mode.elements.mode_tags.elements) do
+      if t.elements.value.value == clusters.RvcRunMode.types.ModeTag.IDLE then
+        tag = clusters.RvcRunMode.types.ModeTag.IDLE
+        break
+      elseif t.elements.value.value == clusters.RvcRunMode.types.ModeTag.CLEANING then
+        tag = clusters.RvcRunMode.types.ModeTag.CLEANING
+        break
+      elseif t.elements.value.value == clusters.RvcRunMode.types.ModeTag.MAPPING then
+        tag = clusters.RvcRunMode.types.ModeTag.MAPPING
+        break
+      end
+    end
```

Comment:
Do we want to return here if the tag isn't valid?
```suggestion
    end
    if tag == 0xFFFF then
      return
    end
```

Reply:
Yes, we have to insert the supported_mode when tag isn't 0xFFFF. I will modify it. 
```lua
if tag ~= 0xFFFF then
    table.insert(supported_modes, mode.elements.label.value)
    table.insert(supported_modes_with_tag, {mode.elements.label.value, tag})
end
```